core/src/main/java/com/linecorp/armeria/common/stream/ByteBufDeframerInput.java

```diff
@@ -133,28 +140,43 @@ public ByteBuf readBytes(int length) {
 
     private ByteBuf readBytesSlow(int length) {
         ByteBuf value = null;
-        int remained = length;
-        for (ByteBuf buf : queue) {
+        int remaining = length;
+        for (final Iterator<ByteBuf> it = queue.iterator(); it.hasNext();) {
+            final ByteBuf buf = it.next();
             if (!buf.isReadable()) {
+                it.remove();
+                buf.release();
                 continue;
             }
 
-            final int readSize = Math.min(remained, buf.readableBytes());
-            if (readSize == length) {
-                return buf.readRetainedSlice(readSize);
-            } else {
-                if (value == null) {
-                    value = alloc.buffer(length);
+            final int readableBytes = buf.readableBytes();
+            if (value == null) {
+                if (readableBytes == length) {
+                    it.remove();
+                    return buf;
                 }
-                value.writeBytes(buf, readSize);
-                remained -= readSize;
-                if (remained == 0) {
-                    break;
+
+                if (readableBytes > length) {
+                    return buf.readRetainedSlice(length);
                 }
+
+                value = alloc.buffer(length);
+            }
+
+            final int readSize = Math.min(remaining, readableBytes);
+            value.writeBytes(buf, readSize);
+            if (readableBytes == readSize) {
+                it.remove();
+                buf.release();
+            }
+
+            remaining -= readSize;
+            if (remaining == 0) {
+                break;
             }
         }
 
-        if (remained > 0 || value == null) {
+        if (remaining > 0 || value == null) {
             ReferenceCountUtil.release(value);
             throw newEndOfInputException();
         }
```

Comment:
We are sure `remaining > 0` here, and thus this could be simplified into:

```java
ReferenceCountUtilk.release(value);
throw newEndOfInputException();
```